### PR TITLE
Financial Connections: for networking manual entry, mid-flow consent, handled extra edge cases like step up.

### PR DESCRIPTION
**Pull request by kgaidis-stripe**
## Summary

[Based on Stripe.js](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/pull/23598)

This PR adds support for when a user in consent screen presses "manual entry" and logs into Link to see RUX/Link Account Picker.

The Link Account Picker will be modified where the button will now say "Agree and connect account" instead of "Connect account." This is because we need to get user consent (they skipped consent by clicking manual entry). 

This also causes a side-effect where the "update required" drawers, which used to be presented on tap, now need to be presented after consent.

So this PR:
1) If consent is required in RUX/Link Account Picker, a drawer will not be presented on tap
2) User consent will be acquired if `acquireConsentOnPrimaryCtaClick`
3) That same drawer will now be presented on click of agree

### This Video Shows The Drawer Not Open On Tap (And Loading Spinner For Consent)

https://github.com/user-attachments/assets/1b55994b-7e98-4614-b3c5-c9eb8ffdf773

### This Video Shows Other Type Of Drawers Still Working

https://github.com/user-attachments/assets/3b69ce00-caaf-4fa6-bea8-f5ad4bc47c3c

### This Video Shows Step Up Verification Working (See Loading Spinner For Acquire Consent)

https://github.com/user-attachments/assets/4f70fa1f-8233-4d69-8de4-680ef1406cdf

image

### All The Existing Other Tests Pass!

```
FinancialConnectionsNetworkingUITests
    ✓ testNativeNetworkingManualEntryTestMode (107.516 seconds)
    ✓ testNativeNetworkingTestMode (133.921 seconds)
    ✓ testNativeNetworkingTestModeSignUpWithMultiSelectAndPrefilledEmail (32.324 seconds)
    ✓ testNativeNetworkingTestModeSignUpWithPrefilledEmailAndPhone (24.578 seconds)
FinancialConnectionsUITests
    ✓ testDataLiveModeOAuthNativeAuthFlow (29.810 seconds)
    ✓ testDataLiveModeOAuthWebAuthFlow (20.255 seconds)
    ✓ testDataTestModeOAuthNativeAuthFlow (20.578 seconds)
    ✓ testPaymentSearchInLiveModeNativeAuthFlow (25.280 seconds)
    ✓ testPaymentTestModeLegacyNativeAuthFlow (19.068 seconds)
    ✓ testPaymentTestModeManualEntryAutofill (13.591 seconds)
    ✓ testPaymentTestModeManualEntryNativeAuthFlow (24.333 seconds)
    ✓ testWebInstantDebitsFlow (13.146 seconds)


	 Executed 12 tests, with 0 failures (0 unexpected) in 464.400 (464.409) seconds
```

### CODE REVIEW

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 1]
```diff
@@ -458,100 +604,15 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
-                                animated: true
-                            )
-                        },
-                        didSelectURL: { [weak self] url in
-                            guard let self = self else { return }
-                            self.didSelectURLInTextFromBackend(url)
-                        },
-                        willDismissSheet: willDismissSheet
-                    )
-                    genericInfoViewController.present(on: self)
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:554]: this logic got moved into `presentAccountUpdateRequiredDrawer`

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 2]
```diff
@@ -297,44 +318,164 @@ final class LinkAccountPickerViewController: UIViewController {
...
+            } else {
+                pushToNextPane()
+            }
+        }
+    }
+
+    // the "account update required drawer" offers the user two choices:
+    // 1. re-link bank account by going through the bank auth flow again (partner_auth pane)
+    // 2. repair the bank account (bank_auth_repair)
+    private func presentAccountUpdateRequiredDrawer(
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:381]: the code here is not new, it just got moved from below

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 3]
```diff
@@ -458,100 +604,15 @@ extension LinkAccountPickerViewController: LinkAccountPickerBodyViewDelegate {
...
-                    let deselectPreviouslySelectedAccount = { [weak self] in
-                        guard let self = self else { return }
-                        self.dataSource.updateSelectedAccounts(
-                            self.dataSource.selectedAccounts.filter(
-                                { $0.partnerAccount.id != selectedPartnerAccount.id }
-                            )
+                    // if we need to acquire consent, instead of showing the drawer now,
+                    // we will show the drawer when user presses the CTA where
+                    // pressing the CTA will make a call to `markConsentAcquired`
+                    if !dataSource.acquireConsentOnPrimaryCtaClick {
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:610]: this is the key part:  IF we need to acquire consent on CTA click, we will NOT present the drawer, here....we will present the drawer after acquiring consent (after CTA click)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 4]
```diff
@@ -297,44 +318,164 @@ final class LinkAccountPickerViewController: UIViewController {
...
+            }
+
+            if dataSource.acquireConsentOnPrimaryCtaClick {
+                footerView?.showLoadingView(true)
+                dataSource
+                    .markConsentAcquired()
+                    .observe { [weak self] _ in
+                        guard let self = self else { return }
+                        footerView?.showLoadingView(false)
+                        pushToNextPane()
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:370]: if `acquireConsentOnPrimaryCtaClick`, only push pane on CTA click

[stripe.js](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/pull/23598/files#diff-96f25fd92e7a6ede67b5de1870616635186357cb3642022af8a41e0c25a968faR376-R389)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 5]
```diff
@@ -268,7 +268,28 @@ final class LinkAccountPickerViewController: UIViewController {
...
+            footerView?.showLoadingView(true)
+            dataSource
+                .markConsentAcquired()
+                .observe { [weak self] _ in
+                    guard let self = self else { return }
+                    footerView?.showLoadingView(false)
+                    self.presentAccountUpdateRequiredDrawer(
+                        drawerOnSelection: drawerOnSelection,
+                        partnerAccount: selectedAccount.partnerAccount
+                    )
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:290]: the logic here is a "reversal" of what happened here:

this code is a follow up from here:

https://github.com/stripe/stripe-ios/pull/3870/files?diff=split&w=1#r1701045857

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 6]
```diff
@@ -268,7 +268,28 @@ final class LinkAccountPickerViewController: UIViewController {
...
+                // to better handle multi-select, we want to ensure that
+                // the account we pick has a `drawerOnSelection`
+                where: { $0.accountPickerAccount.drawerOnSelection != nil }
+            ),
+            let drawerOnSelection = selectedAccount.accountPickerAccount.drawerOnSelection,
+            IsAccountUpdateRequired(forAccount: selectedAccount.partnerAccount)
+        {
+            footerView?.showLoadingView(true)
+            dataSource
+                .markConsentAcquired()
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:283]: [Stripe.js logic](https://git.corp.stripe.com/stripe-internal/stripe-js-v3/pull/23598/files#diff-96f25fd92e7a6ede67b5de1870616635186357cb3642022af8a41e0c25a968faR309-R334)

File: Example/FinancialConnections Example/FinancialConnectionsUITests/FinancialConnectionsNetworkingUITests.swift [hunk 7]
```diff
@@ -459,4 +460,29 @@ final class FinancialConnectionsNetworkingUITests: XCTestCase {
 
         XCTAssert(app.fc_playgroundSuccessAlertView.exists)
     }
+
+    private func executeNativeNetworkingManualEntryTestModeNotNowFlowTest(emailAddress: String) {
```
Comment (kgaidis-stripe) [inline Example/FinancialConnections Example/FinancialConnectionsUITests/FinancialConnectionsNetworkingUITests.swift:464]: this test is unrelated to these specific changes, I just came up with it and wanted to add it

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerDataSource.swift [hunk 8]
```diff
@@ -23,12 +23,14 @@ protocol LinkAccountPickerDataSource: AnyObject {
...
     var analyticsClient: FinancialConnectionsAnalyticsClient { get }
     var dataAccessNotice: FinancialConnectionsDataAccessNotice? { get }
+    var acquireConsentOnPrimaryCtaClick: Bool { get }
 
     func updateSelectedAccounts(_ selectedAccounts: [FinancialConnectionsAccountTuple])
     func fetchNetworkedAccounts() -> Future<FinancialConnectionsNetworkedAccountsResponse>
     func selectNetworkedAccounts(
         _ selectedAccounts: [FinancialConnectionsPartnerAccount]
     ) -> Future<FinancialConnectionsInstitutionList>
+    func markConsentAcquired() -> Promise<FinancialConnectionsSessionManifest>
```
Comment (mats-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerDataSource.swift:33]: Could you help me understand why this API returns a `Promise`, while the others return a `Future`?

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift [hunk 9]
```diff
@@ -615,3 +676,12 @@ private func ZipAccounts(
...
 }
+
+private func IsAccountUpdateRequired(
+    forAccount account: FinancialConnectionsPartnerAccount
+) -> Bool {
+    // repair flow
+    return account.nextPaneOnSelection == .bankAuthRepair
+    // supportability -- account requires re-sharing with additonal permissions
+    || account.nextPaneOnSelection == .partnerAuth
+}
```
Comment (mats-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:687]: To me, this seems like a good candidate for an `extension` on `FinancialConnectionsPartnerAccount`:

```swift
private extension FinancialConnectionsPartnerAccount {
    var updateRequired: Bool {
        // repair flow
        nextPaneOnSelection == .bankAuthRepair
        // supportability -- account requires re-sharing with additional permissions
        || nextPaneOnSelection == .partnerAuth
    }
}
```

And this way it can be called as `account.updateRequired` vs `IsAccountUpdateRequired(forAccount: account)`, which to me is cleaner but it's up to you!

**Review by mats-stripe — APPROVED** [review]
LGTM! Just posted some small questions / suggestions I had 👍

File: Example/FinancialConnections Example/FinancialConnectionsUITests/FinancialConnectionsNetworkingUITests.swift [hunk 10]
```diff
@@ -406,6 +406,8 @@ final class FinancialConnectionsNetworkingUITests: XCTestCase {
         let emailAddresss = "\(UUID().uuidString)@UITestForIOS.com"
         executeNativeNetworkingManualEntryTestModeSignUpFlowTest(emailAddress: emailAddresss)
         executeNativeNetworkingManualEntryTestModeSignInFlowTest(emailAddress: emailAddresss)
+        executeNativeNetworkingManualEntryTestModeConsentWithUpdateRequiredTest(emailAddress: emailAddresss)
+        executeNativeNetworkingManualEntryTestModeNotNowFlowTest(emailAddress: emailAddresss)
```
[the remark this answers is not in the document]
Comment (kgaidis-stripe) [reply Example/FinancialConnections Example/FinancialConnectionsUITests/FinancialConnectionsNetworkingUITests.swift:410]: @mats-stripe 

yea it is not great

there might be another way to structure this

the reason is that for Networking/Link tests we need "state" that is built up to do things (so we need to take actions with a specific, new email AND in a specific order)

maybe there's a better way to accomplish this

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerDataSource.swift (hunk 8 above)
Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerDataSource.swift:33]: @mats-stripe 

this was just a copy-paste that's faulty from old code (consent data source)

it should be a Future

_(there's a longer backstory where the Future/Promise implementation is questionable)_

Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerDataSource.swift]: fixed here, thank you! https://github.com/stripe/stripe-ios/pull/3870/commits/7c988078c4e02fe0a2ac124c22f51914025cec00

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift (hunk 9 above)
Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/LinkAccountPicker/LinkAccountPickerViewController.swift:687]: yea it does look better to have an extension, my main 'worries' are around complexity:
- the logic for "is account update required" is a bit _weird_: it's very specific to this pane (and the proof for this is that I believe this _is_ the only pane where this is checked)
- similarly, one of the things I always try to do is to keep code only where its used (for example, I'll try to avoid creating a function unless its used in multiple places, I'll avoid creating an instance variable unless it needs to be referenced in multiple places etc.); this helps to "reduce complexity" because, by definition, the 'scope' of the code logic is isolated to only certain parts of the code. Of course, for this particular example, its not that much complexity, but if you multiply it by 1000's of decisions where one keeps code isolated to only one place, it starts making a difference in the long-run
